### PR TITLE
KAFKA-2403; Add support for commit metadata in KafkaConsumer

### PR DESCRIPTION
**Pull request by hachikuji**

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java [hunk 1]
```diff
@@ -71,12 +71,12 @@
     /**
      * @see KafkaConsumer#commit(Map, CommitType)
      */
-    public void commit(Map<TopicPartition, Long> offsets, CommitType commitType);
+    public void commit(Map<TopicPartition, OffsetMetadata> offsets, CommitType commitType);
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java]: I feel enforcing users to create OffsetMetadata upon commit brings some overhead, especially in practice most people do not want to embed any metadata with their commits. I would like to propose an alternative solution regarding the API:

commit(Map<TopicPartition, Long>, ComitType, String /\* Commit Message _/)
commit(Map<TopicPartition, Long>, ComitType, String /_ Commit Message */, ConsumerCommitCallBack)

And the commit message will be used as the metadata for all the partitions included. It is based on my assumption that users usually would have a single commit message (i.e. the metadata string) per each commit call; if they want to have different messages for different partitions (for example in CopyCat @ewencp), they can call commit() multiple times with the metadata strings on each partition. This of course may cause more round trips for all sync-commit, but I would suggest people to use the following pattern:

commit(async);
commit(async);
...
commit(sync);  // last call

Since now the commit calls are all ordered.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Coordinator.java [hunk 2]
```diff
@@ -457,38 +461,32 @@ public void handle(OffsetCommitResponse commitResponse, RequestFuture<Void> futu
...
                     log.debug("Committed offset {} for partition {}", offset, tp);
                     if (subscriptions.isAssigned(tp))
                         // update the local cache only if the partition is still assigned
-                        subscriptions.committed(tp, offset);
-                } else if (errorCode == Errors.CONSUMER_COORDINATOR_NOT_AVAILABLE.code()
-                        || errorCode == Errors.NOT_COORDINATOR_FOR_CONSUMER.code()) {
-                    coordinatorDead();
-                    future.raise(Errors.forCode(errorCode));
-                    return;
-                } else if (errorCode == Errors.OFFSET_METADATA_TOO_LARGE.code()
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Coordinator.java:457]: Do we handle these two errors somewhere else?

File: clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java (hunk 1 above)
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java]: @guozhangwang Yeah I agree that it's annoying for most users who don't care about offset metadata. Would the expectation in your suggested API be that the commit message would be tied to the metadata for each topic partition in the map?

Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java]: @guozhangwang when you say overhead, do you mean from an API usability perspective or from a performance perspective?

Comment (ewencp) [inline clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java]: Hmm, not sure I like that API. Having a map for the offsets and then requiring all TPs to use the same metadata is odd. And it seems pretty annoying if you want to do async and get a callback when the commit completes -- now you have to manage waiting on all the callbacks to complete as well. Also, splitting the commit across a bunch of separate requests isn't ideal.

Another alternative is to add this variant:

```
public void commit(Map<TopicPartition, OffsetAndMetadata> offsets, CommitType commitType, ConsumerCommitCallback callback);
```

but leave the existing `Map<TopicPartition, Long>` versions. If you want to include metadata, you _must_ provide all the other parameters, which avoids introducing 2x the number of APIs and is reasonable since using metadata is an advanced feature. This has the same impact as the `String commitMessage` variant since I assume that would be an additional method since most users do not want it. So both would add 1 method, but this version does everything in one round trip, only has one callback, etc.

Copycat actually doesn't have a use case for the metadata currently, so I unfortunately don't have a good example use case for any of this currently.

Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java]: @ewencp To be clear, that would mean we couldn't provide the same version of that method which uses a Long in the Map, right? As a user, I would probably rather have all the commit methods be consistent, but it may be a reasonable compromise.

Comment (ewencp) [inline clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java]: Uggh, type erasure. You're right, we couldn't have both. It's ugly, but we could also use a different name, e.g. `commitWithMetadata`.

Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java]: I've got an even uglier idea. Reverse the order of the last two arguments:

``` java
public void commit(Map<TopicPartition, OffsetAndMetadata> offsets, ConsumerCommitCallback callback, CommitType commitType);
```

More seriously, I don't think commitWithMetadata is so bad.

Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java]: It it really bad to pass OffsetMetadata instead of Long in the map? It adds type safety so is the concern around verbosity, allocation overhead or both?

Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java]: @ijuma I personally don't think it's too bad, but it does add some semantic overhead to the API. Most users will probably just want to commit offsets, but this forces them to deal with the notion of associated metadata (even if they can just leave it empty).

Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java]: You could call the class Offset (since the metadata is just an optional field).

Comment (ewencp) [inline clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java]: I don't think it's _that_ big a deal to have to allocate the `OffsetMetadata`. And certainly the performance overhead of the allocation isn't a concern. I only care about the verbosity because the vast majority of use cases only care about the offset and not the metadata, and we're making that large fraction of cases harder. And would OffsetMetadata then be changed to be mutable, so it's convenient to just maintain the map where I update only the offset in that struct? Or do all my updates to that map (which I probably update for every single message processed) require a `new OffsetMetadata()`, bloating those statements and making them less clear? Or do I just maintain the `Map<TopicPartition, OffsetMetadata>` and have to convert it every time I call commit?

On the other hand, maybe most users don't even specify the offsets manually anyway and the concern here is unwarranted since 99% of the cases are handled by `commit(CommitType)` and `commit(CommitType, ConsumerCommitCallback)`?

In other words, I'm worried because I want the very common case to be clean, easy to read, and concise. I'm not yet sure whether this change would actually affect that common case.

Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java]: As an alternative, I was also considering the following:

``` java
void commit(TopicPartition partition, Long offset, String metadata);
```

That's annoying when you need to commit metadata for multiple partitions, especially if you need to provide a callback. Of the options so far, my preference is for `commitWithMetadata`.

Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java]: @ewencp As you say, it does hinge on whether this `commit` overload is used often or not. If it is not, then having this _and_ having `commitWithMetadata` seems excessive. You guys have a better handle on this, so I'll leave it to you. :)

Comment (nehanarkhede) [inline clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java]: > On the other hand, maybe most users don't even specify the offsets manually anyway and the concern here is unwarranted since 99% of the cases are handled by commit(CommitType) and commit(CommitType, ConsumerCommitCallback)

Agree with @ewencp that manual offset commit is really a very small percentage of all uses. Even though I agree that amongst that minority, fewer would have custom metadata, I'm not sure it is worth adding the extra commitWithMetadata API for. 

It may be ok in this case to go with 
`public void commit(Map<TopicPartition, OffsetMetadata> offsets, CommitType commitType);`

Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java]: +1

Comment (jkreps) [inline clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java]: The purpose of the `Map<TopicPartition, Long>` was to avoid adding a new object type. But since we're doing that anyway what about just making the call:

```
public void commit(List<OffsetMetadata> offsets, CommitType type)
```

where now the `OffsetMetadata` class includes the `TopicPartition`? This is arguably no more complex than the original call in the case where you aren't giving metadata.

Also does `OffsetMetadata` imply metadata about the offset whereas in fact this is both the offset and metadata? Other options would be `OffsetCommit` or `OffsetInfo` or `PartitionOffset`. Don't have a strong feeling on this one.

Comment (ewencp) [inline clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java]: The problem with the `List` variant is that it's inconvenient for the application to store it that way -- it's most convenient if the API uses a layout that works well for applications, and since the application probably needs to do lookups by `TopicPartition`, using a `Map` is natural. Also, the map guarantees correct semantics, i.e. only one entry per `TopicPartition`.

I think requiring `OffsetAndMetadata` or whatever name you want to use is fine as long as it's mutable. It's a little overhead for people who only want to set the offset manually, but it's tiny and if its mutable there's not much to complain about wrt object allocation overhead or anything like that.

Also, I realized that the ownership requirements for the data associated with async commits is probably not clear from the javadocs. i.e. do I need to make sure the data I passed in doesn't change until the commit callback is invoked, or is KafkaConsumer going to ensure the data is copied so I'm free to start modifying the map I pass in as soon as the commit method returns?

Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java]: @ewencp List might not be great, but Collection might work. The user could still use a Map, and then pass map.values() to commit. 

Also, that's a good point on mutability for the async case. I'd rather not allow that data to change while the commit is still pending, so maybe an internal copy is the best answer?

Comment (ewencp) [inline clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java]: Not sure how using Collection would really help -- the element type for Maps would be `Entry<TopicPartition, OffsetAndMetadata>`, which won't implement any useful interface that would be interchangeable with a `List<OffsetCommitInfo>` where the `OffsetCommitInfo` contains both the topic partition and the offset and metadata.

For mutability, in Copycat I just handled this at the application level -- subsequent updates are handled in a secondary map and they are merged once the commit completes. Would be nice to avoid having to do this if possible.

Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java]: I was thinking that the user would maintain `Map<TopicPartition, OffsetMetadata>`, so map.values() would return `Collection<OffsetMetadata>`, which could be passed directly to commit. That said, I don't have a strong preference on any of the choices so far, but I'm probably leaning toward the initial approach. The overhead is not so high and users who want to manage their own commits may find metadata more useful than we're imagining. Off the top of my head, I can imagine wanting to add timestamps, hostnames, or maybe versioning information to the commit. If instead it's really such an extreme edge case that almost no users would find value in, maybe it doesn't belong in the API at all and we can think about another way to expose it.

Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java]: @ewencp Regarding making `OffsetAndMetadata` mutable, does that make sense if we do defensive copies of the passed in map as is being suggested? It seems to me that ownership rules would be inconsistent in that case and we would be doing plenty of allocation with the map copy anyway. It's also worth saying that due to boxing, we were already allocating `Long` instances, so the number of allocations doesn't change, it's just the size of each instance.

Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java]: OK, now I think my proposed approach does not suits well as well after reading through the conversations. If we agree that the percentage of users calling commit with manual offset values is really low, then I think it may be OK for adding this API usability overhead for them with a probably even lower percentage usage. So let's just proceed with the original API.

Comment (nehanarkhede) [inline clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java]: > So let's just proceed with the original API.
> @guozhangwang Can you clarify?

Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java]: @nehanarkhede Yup I can review and check it in once it is rebased.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Coordinator.java (hunk 2 above)
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Coordinator.java:457]: I just condensed the error checking here, since all we were doing is logging the errors. We'll still log them in the updated patch, but they didn't seem to require their own branch.

File: clients/src/main/java/org/apache/kafka/clients/consumer/OffsetAndMetadata.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+
+        OffsetAndMetadata that = (OffsetAndMetadata) o;
+
+        if (offset != that.offset) return false;
+        return !(metadata != null ? !metadata.equals(that.metadata) : that.metadata != null);
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/OffsetAndMetadata.java]: How about "metadata == null ? that.metadata == null : metadata.equals(that.metadata)"?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/SubscriptionStateTest.java [hunk 4]
```diff
@@ -143,7 +153,7 @@ public void cantChangeConsumedPositionForNonAssignedPartition() {
     }
     
     public void assertAllPositions(TopicPartition tp, Long offset) {
-        assertEquals(offset, state.committed(tp));
+        assertEquals(offset, (Long) state.committed(tp).offset());
```
Comment (guozhangwang) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/SubscriptionStateTest.java]: Do we need this casting?

Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/SubscriptionStateTest.java]: Annoyingly, yes. We wouldn't need it if we changed the offset field in OffsetAndMetadata to use a Long instead of long.

Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/SubscriptionStateTest.java]: A better way is to do `offset.longValue` on the left side.

Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/SubscriptionStateTest.java]: Good call.

**Comment by hachikuji** [comment]
@guozhangwang Made a few minor changes to the comments and added a defensive copy of the passed map in commitAsync.

**Comment by guozhangwang** [comment]
LGTM.
